napari/utils/context/_expressions.py

```diff
@@ -0,0 +1,597 @@
+                kwargs[name] = [self.visit(item) for item in field]
+            else:
+                kwargs[name] = field
+        return globals()[type_](**kwargs)
+
+    # can drop after py3.7 support is gone
+    _PY37_CONSTS = {
+        'Num': 'n',
+        'Str': 's',
+        'Bytes': 's',
+        'NameConstant': 'value',
+    }
+
+
+class ExprSerializer(ast.NodeVisitor):
+    """Serializes an :class:`Expr` into a string.
+
+    Examples
+    --------
+    >>> expr = Expr.parse('a + b == c')
+    >>> print(expr)
+    'a + b == c'
+
+    or ... using this visitor directly:
+
+    >>> ser = ExprSerializer()
+    >>> ser.visit(expr)
+    >>> out = "".join(ser.result)
+    """
+
+    def __init__(self, node: Optional[Expr] = None) -> None:
+        self._result: List[str] = []
+
+        def write(*params):
+            for item in params:
+                if isinstance(item, ast.AST):
+                    self.visit(item)
+                elif item:
+                    self._result.append(item)
+
+        self.write = write
+
+        if node is not None:
+            self.visit(node)
+
+    def __str__(self):
+        return "".join(self._result)
+
+    def visit_Name(self, node: ast.Name) -> None:
+        self.write(node.id)
+
+    def visit_Constant(self, node: ast.Constant) -> None:
+        self.write(repr(node.value))
+
+    def visit_BoolOp(self, node: ast.BoolOp) -> Any:
+        op = f' {_OPS[type(node.op)]} '
+        for idx, value in enumerate(node.values):
+            self.write(idx and op or '', value)
+
+    def visit_Compare(self, node: ast.Compare) -> None:
+        self.visit(node.left)
+        for op, right in zip(node.ops, node.comparators):
+            self.write(f' {_OPS[type(op)]} ', right)
+
+    def visit_BinOp(self, node: ast.BinOp) -> None:
+        self.write(node.left, f' {_OPS[type(node.op)]} ', node.right)
+
+    def visit_UnaryOp(self, node: ast.UnaryOp) -> None:
+        sym = _OPS[type(node.op)]
+        self.write(sym, ' ' if sym.isalpha() else '', node.operand)
+
+    def visit_IfExp(self, node: ast.IfExp) -> Any:
+        self.write(node.body, ' if ', node.test, ' else ', node.orelse)
+
+
+OpType = Union[
+    Type[ast.operator], Type[ast.cmpop], Type[ast.boolop], Type[ast.unaryop]
+]
+_OPS: Dict[OpType, str] = {
+    # ast.boolop
+    ast.Or: 'or',
+    ast.And: 'and',
+    # ast.cmpop
+    ast.Eq: '==',
+    ast.Gt: '>',
+    ast.GtE: '>=',
+    ast.In: 'in',
+    ast.Is: 'is',
+    ast.NotEq: '!=',
+    ast.Lt: '<',
+    ast.LtE: '<=',
+    ast.NotIn: 'not in',
+    ast.IsNot: 'is not',
+    # ast.operator
+    ast.BitOr: '|',
+    ast.BitXor: '^',
+    ast.BitAnd: '&',
+    ast.LShift: '<<',
+    ast.RShift: '>>',
+    ast.Add: '+',
+    ast.Sub: '-',
+    ast.Mult: '*',
+    ast.Div: '/',
+    ast.Mod: '%',
+    ast.FloorDiv: '//',
+    ast.MatMult: '@',
+    ast.Pow: '**',
+    # ast.unaryop
+    ast.Not: 'not',
+    ast.Invert: '~',
+    ast.UAdd: '+',
+    ast.USub: '-',
+}
```

Comment:
I'd love to fuzz that with hypothesis, and make sure the (parse->unparse-> parse) and (unparse->parse-> unparse) are the identity (to some normalisation to the string form). But looks great.

Reply:
yeah, agree.  The closes thing I have at the moment are these round trip tests [here](https://github.com/napari/napari/pull/3350/files#diff-8f34b9e8c0505b9f4e704afaca1d1a3bc4d650e25c7f7943b0609c97eaf52a74R191-R192)... but I suspect that I'm missing many cases